### PR TITLE
Refresh the HA guide

### LINKED ISSUE #22742: Refresh our general high availability guide
## Applies To
https://github.com/gravitational/teleport/blob/master/docs/pages/deploy-a-cluster/high-availability.mdx

## Details

As of v12, the `teleport-cluster` Helm chart deploys the Auth Service and Proxy Service separately, using an external load balancer for the Proxy Service and an internal load balancer for the Auth Service. This is also the architectural approach we recommend for production Teleport deployments. Let's update the guide—including the diagram—to match this approach.

### PR DESCRIPTION
**Pull request by ptgott**
Closes #22742

Update the HA guide to reflect v12 changes in the `teleport-cluster` Helm chart, e.g., running the Auth Service and Proxy Service in separate compute pools.

### CODE REVIEW

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 1]
```diff
@@ -16,14 +17,17 @@ deployment.
...
-
-- **A Layer 4 load balancer** to direct traffic from users and services to an
-  available `teleport` process.
+A high-availability Teleport cluster revolves around two pools of redundant
+`teleport` processes, one running the Auth Service and one running the Proxy
+Service, plus the infrastructure required to support each pool. 
+
+Infrastructure components include:
+- A **public Layer 4 load balancer** to direct traffic from users and services
+  to an available `teleport` process.
```
Comment (zmb3) [inline docs/pages/deploy-a-cluster/high-availability.mdx:26]:
```suggestion
  to an available Teleport Proxy Service.
```

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 2]
```diff
@@ -44,21 +48,34 @@ This includes:
...
+  an available Proxy Service instance. This load balancer handles TCP traffic
+  from users and services in a variety of application-layer protocols.
+- **Auth Service load balancer:** A load balancer to forward traffic from a
+  Proxy Service instance to an available Auth Service instance. This handles TLS
+  traffic to the Auth Service's gRPC endpoint. It can also handle traffic to
+  Teleport Kubernetes Service instances running in the Auth Service pool in
+  order to enable secure access to your Teleport cluster's infrastructure.
 
-We recommend configuring your load balancer to route traffic across multiple
+Neither load balancer must terminate TLS. Both load balancers must transparently
```
Comment (zmb3) [inline docs/pages/deploy-a-cluster/high-availability.mdx:64]: "Neither load balancer must" makes it sound like terminating TLS is optional and the LB doesn't have to, but can, terminate TLS.

Should we be stricter here and say "You must not terminate TLS on either of these load balancers"??

(That said, we technically do support ALBs for most protocols now..)

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 3]
```diff
@@ -96,7 +114,7 @@ These ports are required:
 
 | Port | Description |
 | - | - |
-| `3023` | SSH port for clients connect to. |
+| <nobr>`3023`</nobr> | SSH port for clients connect to. |
 | `3024` | SSH port used to create reverse SSH tunnels from behind-firewall environments. |
 | `443` | HTTPS connections to authenticate `tsh` users into the cluster. The same connection is used to serve a Web UI. |
```
Comment (zmb3) [inline docs/pages/deploy-a-cluster/high-availability.mdx:119]: Do we specifically tell the user to configure the Teleport proxy to listen on port 443 in this guide?

Would it make more sense to list both the frontend and backend ports here, and have the LB direct incoming traffic on port 443 to port 3080 on the proxy? (Teleport would require additional privileges to bind to 443 as opposed to 3080)

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 4]
```diff
@@ -112,6 +130,19 @@ services:
...
 </Tabs>
 
+### Configuring the Auth Service load balancer
+
+The Auth Service load balancer must forward traffic to the Auth Service's gRPC
+port. In this guide, we are assuming that you have configured the Auth Service
+load balancer to forward traffic from port `3025` to port `3025` on an available
+Auth Service instance. 
+
+If you are running your Teleport cluster on Kubernetes, you can configure your
```
Comment (zmb3) [inline docs/pages/deploy-a-cluster/high-availability.mdx:140]: Hmm, this confuses me a bit. Auth never listens on port 3026.

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 5]
```diff
@@ -256,23 +287,30 @@ records.
 
 ## Teleport instances
 
-Run the Teleport Auth Service and Proxy Service as a scalable group of compute
-resources, for example, a Kubernetes `Deployment` or AWS Auto Scaling group.
-This requires running the `teleport` binary on each Kubernetes pod or virtual
-machine or in your group. 
+Run the Teleport Auth Service and Proxy Service as two scalable groups of
+compute resources, for example, using Kubernetes `Deployment`s or AWS Auto
```
Comment (zmb3) [inline docs/pages/deploy-a-cluster/high-availability.mdx:291]:
```suggestion
compute resources, for example, using Kubernetes Deployments or AWS Auto
```

I don't think we need to render the word deployments in a code block.

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 6]
```diff
@@ -451,6 +442,111 @@ should not have direct access to the underlying node.
...
+### Auth Service pool
+
+#### Open ports
+
+Ensure that, on each Auth Service instance, the following ports are open: 
+
+| Port | Description |
+| - | - |
+| <nobr>`3025`</nobr> | gRPC port to open to Proxy Service instances.|
+| `3026` | Optional port to open for traffic to the Teleport Kubernetes Service (if deploying on Kubernetes).|
```
Comment (zmb3) [inline docs/pages/deploy-a-cluster/high-availability.mdx:454]: I don't understand this. By this logic, should we also add the desktop service port (3028) because you can deploy desktop service on auth instances as well?

(That's a rhetorical question - I believe the answer is no, I just don't understand why kubernetes_service is getting special treatment here)

File: docs/pages/deploy-a-cluster/high-availability.mdx (hunk 3 above)
Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx:119]: Looks like there's no `web_listen_addr` example here. I'll change the LB instructions!

File: docs/pages/deploy-a-cluster/high-availability.mdx (hunk 4 above)
Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx:140]: I'm basing this on the `teleport-cluster` chart, which enables the Kubernetes Service on Auth Service pods: https://github.com/gravitational/teleport/blob/64448b1b0ed36643aee0bf471171ba677548f4aa/examples/chart/teleport-cluster/templates/auth/_config.common.tpl#L5-L8

File: docs/pages/deploy-a-cluster/high-availability.mdx (hunk 6 above)
Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx:454]: As above, this is based on the `teleport-cluster` chart, where Auth Service pods run the Kubernetes Service. My understanding is that Auth Service pods would run the Kubernetes Service to enable access to (at least) Teleport pods running on Kubernetes. Is that accurate, @hugoShaka ?

Comment (hugoShaka) [reply docs/pages/deploy-a-cluster/high-availability.mdx:454]: Yeah, it's a bit complex and specific to the Helm chart. We packaged kube access at some point and have to live with this. We also cannot connect kube access to the proxy because we can't establish trust with the proxy without relying on external infra (LE, DNS, LoadBalancer) (discussed [here](https://github.com/gravitational/teleport/issues/19275#issuecomment-1477978729)). I tried to extract this service but failed, and users will be upset if we remove it.

tl;dr I would remove mention of the kube access port in a generic HA guide, this is an implementation detail because we made questionable choices

Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx]: 👍 I've removed the Kubernetes mentions

File: docs/pages/deploy-a-cluster/high-availability.mdx (hunk 2 above)
Comment (zmb3) [reply docs/pages/deploy-a-cluster/high-availability.mdx:62]: Paging @greedy52 who can answer this better than I can

Comment (hugoShaka) [reply docs/pages/deploy-a-cluster/high-availability.mdx:62]: Afaik:
- Proxy -> Auth LB must not terminate TLS
- Client -> Proxy LB can terminate TLS since 13.0 (but this might come with additional restrictions for kube access)

Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx:62]: Thanks! What additional restrictions are these? Is there anything else I should know,  @greedy52?

Comment (greedy52) [reply docs/pages/deploy-a-cluster/high-availability.mdx:62]: As mentioned, it doesn't support all protocols yet and it also has some performance penalties which we may do some performance improvement in v14. So I think layer 4 LBs should still be recommended way to set up HA.

I was planning to write up a separate page for TLS Routing behind layer 7 LB preview. Maybe I can add a note here to link to that page when it's available?

Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx:62]: Sounds good! I'll keep the recommendation the way it is for now and we can add the note when the L7 guide is ready.

**Comment by ptgott** [comment]
@zmb3 Checking to see if there is more work I should be doing on this. It looks like @greedy52 will be adding [separate docs](https://github.com/gravitational/teleport/pull/24479#discussion_r1173247169) for L7 LBs, and I've removed the Kubernetes-specific language. 

Also paging @alexfornuto for a copy review.

Thanks!

File: docs/img/deploy-a-cluster/teleport-ha-architecture.png
Comment (alexfornuto) [inline docs/img/deploy-a-cluster/teleport-ha-architecture.png:1]: I suggest increasing the size of the "Developers" icon and text, I'm unable to read it when scaled into the page at my standard viewing (half a 1440 screen).

File: docs/pages/deploy-a-cluster/high-availability.mdx (hunk 2 above)
Comment (ptgott) [reply docs/pages/deploy-a-cluster/high-availability.mdx:62]: I think "Both..must not" is ambiguous, since it can also mean "one can". I've suggested: 

```
Both load balancers must transparently forward the TCP traffic they receive,
without terminating TLS. In other words, these must be Layer 4 load balances,
not Layer 7 (e.g., HTTP) load balancers.
```

File: docs/img/deploy-a-cluster/teleport-ha-architecture.png
Comment (ptgott) [reply docs/img/deploy-a-cluster/teleport-ha-architecture.png:1]: Since we still have some time in the quarter, I think it would make sense to replace this with a Mermaid diagram. We would need to get this PR approved, though: https://github.com/gravitational/teleport/pull/25173

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 7]
```diff
@@ -363,17 +361,11 @@ Teleport configuration:
...
 auth_service:
-  enabled: true
-  cluster_name: "mycluster.example.com"
-  # Remove this if not using Teleport Enterprise
-  license_file: "/var/lib/license/license.pem"
+  enabled: false
 proxy_service:
   enabled: true
-  public_addr: "mycluster.example.com:443"
+  public_addr: "mycluster.example.com:3080"
```
Comment (greedy52) [inline docs/pages/deploy-a-cluster/high-availability.mdx:368]:
```suggestion
  public_addr: "mycluster.example.com:443"
```
for what load balancer serves.

File: docs/pages/deploy-a-cluster/high-availability.mdx [hunk 8]
```diff
@@ -390,20 +382,14 @@ Teleport configuration:
...
-  cluster_name: "mycluster.example.com"
-  # Remove this if not using Teleport Enterprise
-  license_file: "/var/lib/license/license.pem"
+  enabled: false
 proxy_service:
   enabled: true
   listen_addr: 0.0.0.0:3023
   tunnel_listen_addr: 0.0.0.0:3024
-  public_addr: "mycluster.example.com:443"
+  public_addr: "mycluster.example.com:3080"
```
Comment (greedy52) [inline docs/pages/deploy-a-cluster/high-availability.mdx:392]: sorry i think there is one more here:
```suggestion
  public_addr: "mycluster.example.com:443"
```

**Review by greedy52 — APPROVED** [review]

**Review by alexfornuto — APPROVED** [review]
